# Notebook 6: Bayes' Rule and the Law of Total Probability
***

In this notebook we'll get some more practice with conditional probabilities, total probability, the product rule, and now Bayes' Rule.  We'll also see how we can do some simple random simulations using Numpy to verify our results. 

We'll need Numpy for this notebook, so let's load it. 

In [1]:
import numpy as np 

### Exercise 1: Bayes' Rule and The Three Doors Problem 
***

Suppose you're on a game show, and you're given the choice of three doors: Behind one door is a car; behind the others, goats. You pick a door, say No. 1, and the host, Monty, who knows what's behind the doors, opens another door, say No. 3, which has a goat. He then says to you, "Do you want to pick door No. 2?" Is it to your advantage to switch your choice?

![doors](https://cdn-images-1.medium.com/max/1600/1*fSv7k4vXkOYp8RN7lVeKyA.jpeg)

**Historical notes:** This was never how "Let's Make a Deal" with Monty Hall was played. ([Here](https://www.youtube.com/watch?v=c1BSkquWkDo) is a snippet from an interview with Monty Hall about it.)  It is a problem posed by Marilyn vos Savant in _Parade_ magazine in 1990.  The fallout was intense.  Read more [here](https://priceonomics.com/the-time-everyone-corrected-the-worlds-smartest/) and [here](http://marilynvossavant.com/game-show-problem/).

**Part A**: What does your intuition say?  Is it in your best interest to switch? 

***
**Solution**:  It is in your interest to switch.  It turns out that if you don't switch doors you have a 1/3 probability of winning the car.  If you do switch doors you have a 2/3 probability.  We'll prove this using Bayes rule in the next part, but if this seems sketchy to you, consider a modified game where there are 100 doors with 99 goats and one car.  After making your guess the host (knowing where the car is) opens 98 doors that have goats behind them.  Now does your intuition change? 
***

**Part B**: Let's assume that you pick door number 1 and Monty opens door number 3.  The question then is whether you stick with door number 1 or switch to door number 2.

Let $\color{red}{D_i}$ be the event that the car is actually behind door $i$.

Let $\color{red}H$ be the event that Monty opens door number 3.

First, compute $P(H \mid D_i)$ for $i=1,2,3$.

You'll need to think carefully about the particular situation described above.

***
**Solution**: The likelihoods, given our assumptions, are as follows: 

- $D_1$: If the car is behind Door 1 then it doesn't matter which of Doors 2 and 3 that Monte opens.  Thus this probability is $P(H \mid D_1) = \frac{1}{2}$

- $D_2$: If the car is behind Door 2 then Monte must open Door 3.  Thus this probability is $P(H \mid D_2) = 1$

- $D_3$: If the car is behind Door 3 then there is no way that Monte will open Door 3.  Thus this probability is $P(H \mid D_3) = 0$

***

**Part C**: Use your results from **Part B** and the Law of Total Probability to compute $P(H)$

***
**Solution**: Assuming that it is equally likely that the car is behind any door, we have $P(D_i) = \frac{1}{3}$ for $i=1, 2, 3$. We then have 

\begin{eqnarray}
\nonumber P(H) &=& P(H \mid D_1)P(D_1) + P(H \mid D_2)P(D_2) + P(H \mid D_3)P(D_3) \\
\nonumber      &=& \frac{1}{2}\cdot\frac{1}{3} + 1\cdot\frac{1}{3} + 0\cdot\frac{1}{3} \\ 
\nonumber      &=& \frac{1}{2}
\end{eqnarray}

***

**Part D**: Now, use Bayes' Rule to compute $P(D_i \mid H)$ for $i=1$ and $2$ (because these are the doors we care about). 

***
**Solution**: From Bayes' Rule, we have 

$$
p(D_1 ~|~ H) = \frac{P(H \mid D_1)P(D_1)}{P(H)} = \frac{\frac{1}{2} \cdot \frac{1}{3}}{\frac{1}{2}} = \frac{1}{3} ~~~~~~~~ \textrm{and} $$

$$~~~~~~~~p(D_2 ~|~ H) = \frac{P(H \mid D_2)P(D_2)}{P(H)}  = \frac{1 \cdot \frac{1}{3}}{\frac{1}{2}} = \frac{2}{3}
$$

Thus if we switch we have a probability of $\frac{2}{3}$ of winning the car. 

***

**Part E**: Write some simple code to simulate the Three Doors problem and verify your results from **Parts A-D**. 

In [2]:
def make_a_deal(switch):
    # This function will return whether or not you won the car,
    #  based on your decision to switch or not.
    
    doors = [1,2,3]
    # a list holding the three door numbers
    
    car = np.random.choice(doors)
    # an array holding the door with the car
    
    first_pick = np.random.choice(doors)
    # an array holding your first pick, before being given
    #  the option to switch.
    
    monty_choices = [n for n in doors if n != first_pick]
    # an array holding monty's door choices. These are
    #   the doors remaining from the one you DIDN'T pick.
    
    if car == first_pick:
        # if the car is behind the door of your first pick,
        #  then enter this if-block
        monty_pick = np.random.choice(monty_choices)
        # inside this if-block 'monty_pick' will contain
        #  a random choice of the remaining two doors.
    else:
        # If the car is not behind the door of your first choice,
        #  then Monty has only one door and has to pick it.
        monty_choices_2 = [n for n in monty_choices if n != car]
        # Do you see that this array holds only ONE door number?
        monty_pick = monty_choices_2[0]
        # Now, put that SINGLE door into 'monty_pick'.
    
    if switch:
        # Enter this if-block if you decide to switch doors.
        final = [n for n in doors if n != monty_pick and n != first_pick]
        # This is the only door left to switch to.
        final_choice = final[0]  
        # This variable holds your switch-to choice door number.
    else:
        final_choice = first_pick
        # If you don't switch, then your final choice is
        #  the same as your first pick.
        
    return final_choice == car
        # The return is a boolean stating whether or not
        #  you got the car door.



def three_doors_sim(switch, num_trials): 
    # This is a funciton that takes as input your decision to
    #  switch or not; True or False, and the number of times you
    #   want to run the simulation for your choice to switch or not.
    
    winners = np.sum([make_a_deal(switch) for kk in range(num_trials)])
    # 'winners' counts the number of times you win by counting the
    #  1's delivered from the 'make_a_deal' function for each trial.
                      
    if switch:
        # Enter the if-block if switch=True, and calculate ratio.
        print("P(winning by switching = {:.4f})".format(winners/num_trials))
    else:
        # Enter the Else-block if switch=False, and calculate ratio.
        print("P(winning by not switching = {:.4f})".format(winners/num_trials))

# Now call the function 10000 times with your choice to keep the door,
#  or to switch to the other door.
three_doors_sim(switch=True, num_trials=int(1e5))
three_doors_sim(switch=False, num_trials=int(1e5))


P(winning by switching = 0.6678)
P(winning by not switching = 0.3326)


In [3]:
# This code is an alternate solution. You can remove the hash-tags in the left column and try it if you like.
#def make_a_deal(switch=True):
#    doors = list(range(3))
#    car = np.random.choice(doors)
#    first_choice = np.random.choice(doors)
#    
#    montes_options = list(set(doors)-set([car])-set([first_choice]))
#    goat = np.random.choice(montes_options)
#    final_choice = (set(doors)-set([goat])-set([first_choice])).pop() if switch else first_choice
#    return final_choice == car
#
#def three_doors_sim(switch=True, num_trials=int(1e3)): 
#    winners = np.sum([make_a_deal(switch) for kk in range(num_trials)])
#    state = "switching" if switch else "not switching"
#    print("P(winning by "+state+" = {:.4f}".format(winners/num_trials))

In [4]:
#three_doors_sim(switch=True, num_trials=int(1e5))

In [5]:
#three_doors_sim(switch=False, num_trials=int(1e5))

### Exercise 2:  What if there are Scooby-Doo numbers of doors?
***

![moredoors](https://thesaurus.plus/img/sentences/39/Scooby-Dooby_Doors.png)

**Part A:** Now suppose there are 4 (or maybe in general, $n$) doors.  There is still only one prize, but now $n-1$ goats, AND Monty is only going to open a single other door. Is it still beneficial to switch with this minimal extra information?

Modify your simulation from above to estimate the probability of winning if you switch to a random other available door, or if you stick with your original choice.

In [6]:
def scooby_doo(ndoors,switch):
    
    #generalized to ndoors. We are still randomly selecting the
    #  car and the contestant's first pick from the door list.
    
    ndoors = list(range(ndoors))
    # so, if ndoors=4, then the variable contains list [0,1,2,3]
    
    car = np.random.choice(ndoors)
    # 'car' contains a randomly chosen door to contain the car.
    
    first_pick = np.random.choice(ndoors)
    # This array holds your first choice of door before the
    #  option to switch to the remaining door.
    
    monty_choices = [n for n in ndoors if n != first_pick]
    # This is an array that contains the remaining doors, 
    #  i.e. not your first pick, nor the door with the car.
    
    if car == first_pick:
        # Enter this if-block if your first  pick actually
        #  contains the car.
        monty_pick = np.random.choice(monty_choices)
        # This variable contains a random choice of a SINGLE door 
        #  chosen from the remaining doors.
    else:
        monty_choices_2 = [n for n in monty_choices if n != car]
        #Significant change here - Monty will choose randomly from
        #  n-2 doors (prior function has him picking the only
        #  remaining element from the list.)

        monty_pick = np.random.choice(monty_choices_2)
        # so, if you chose door 0, and the car was behind door 1,
        #  then Monty will randomly choose ONE door, either 2 or 3.
        
    if switch:
        final = [n for n in ndoors if n != monty_pick and n != first_pick]
        # Now there are n-2 doors for the contestant to
        #  choose from, for their agreed upon switch.
        final_choice = np.random.choice(final)  
    else:
        final_choice = first_pick
        
    return final_choice == car

def n_doors_sim(n, switch, num_trials): 
    
    # True = 1 and false = 0
    winners = np.sum([scooby_doo(n, switch) for kk in range(num_trials)])
    # 'winners' contains the sum of the number of times you win,
    #  according to your decision to switch or not.
    
    if switch:
        print("P(winning by switching = {:.4f})".format(winners/num_trials))   
        # This will print if you chose to switch.
    else:
        print("P(winning by not switching = {:.4f})".format(winners/num_trials))
        # This will print if you chose NOT to swith.

# Now, lets call the function with 10000 trials,
#  and print what happened with 3 doors, 4 doors, 5 doors,...       
print("With 3 doors:")
n_doors_sim(3, switch=True, num_trials=int(1e5))
n_doors_sim(3, switch=False, num_trials=int(1e5))

print("\n With 4 doors")
n_doors_sim(4, switch=True, num_trials=int(1e5))
n_doors_sim(4, switch=False, num_trials=int(1e5))

print("\n With 5 doors")
n_doors_sim(5, switch=True, num_trials=int(1e5))
n_doors_sim(5, switch=False, num_trials=int(1e5))


With 3 doors:
P(winning by switching = 0.6697)
P(winning by not switching = 0.3340)

 With 4 doors
P(winning by switching = 0.3750)
P(winning by not switching = 0.2494)

 With 5 doors
P(winning by switching = 0.2680)
P(winning by not switching = 0.2003)


**Part A - Intuition**

Now suppose there are again more than 3 doors and that Monty opens all doors but one. This is exactly what he does when there are 3 doors. How does probability change when there are more than 3 doors?

In [7]:
def intuition(ndoors,switch):
    
    #generalized to ndoors. We are still randomly selecting the car and the 
    # contestant's first pick from the door list.
    
    ndoors = list(range(ndoors))
    car = np.random.choice(ndoors)
    first_pick = np.random.choice(ndoors)
    monty_choices = [n for n in ndoors if n != first_pick]

    if car == first_pick:
        monty_leaves_closed = np.random.choice(list(set(ndoors) - set([first_pick])))
        
    else:
        monty_leaves_closed = np.random.choice(list(set([car])))
        
    if switch:
        final_choice = monty_leaves_closed
    
    else:
        final_choice = first_pick
    return final_choice == car


def n_doors_sim(n, switch, num_trials): 
    
    # True = 1 and false = 0
    winners = np.sum([intuition(n, switch) for kk in range(num_trials)])
                      
    if switch:
        print("P(winning by switching = {:.4f})".format(winners/num_trials))   
    else:
        print("P(winning by not switching = {:.4f})".format(winners/num_trials))

print("With 3 doors:")
n_doors_sim(3, switch=True, num_trials=int(1e5))
n_doors_sim(3, switch=False, num_trials=int(1e5))
print('')

print("With 4 doors:")
n_doors_sim(4, switch=True, num_trials=int(1e5))
n_doors_sim(4, switch=False, num_trials=int(1e5))
print('')

print("With 5 doors:")
n_doors_sim(5, switch=True, num_trials=int(1e5))
n_doors_sim(5, switch=False, num_trials=int(1e5))
print('')

print("With 6 doors:")
n_doors_sim(6, switch=True, num_trials=int(1e5))
n_doors_sim(6, switch=False, num_trials=int(1e5))
print('')

print("With 7 doors:")
n_doors_sim(7, switch=True, num_trials=int(1e5))
n_doors_sim(7, switch=False, num_trials=int(1e5))
print('')

print("With 100 doors:")
n_doors_sim(100, switch=True, num_trials=int(1e5))
n_doors_sim(100, switch=False, num_trials=int(1e5))


With 3 doors:
P(winning by switching = 0.6678)
P(winning by not switching = 0.3350)

With 4 doors:
P(winning by switching = 0.7527)
P(winning by not switching = 0.2493)

With 5 doors:
P(winning by switching = 0.7979)
P(winning by not switching = 0.1992)

With 6 doors:
P(winning by switching = 0.8342)
P(winning by not switching = 0.1662)

With 7 doors:
P(winning by switching = 0.8574)
P(winning by not switching = 0.1417)

With 100 doors:
P(winning by switching = 0.9895)
P(winning by not switching = 0.0101)


In [8]:
# This again is alternate shorter, more complex code if you want to attempt it.
#def make_a_deal(n, switch=True):
#    doors = list(range(n))
#    car = np.random.choice(doors)
#    first_choice = np.random.choice(doors)
#    montes_options = list(set(doors)-set([car])-set([first_choice]))
#    goat = np.random.choice(montes_options)
#    final_choice = (set(doors)-set([goat])-set([first_choice])).pop() if switch else first_choice
#    return final_choice == car

#def n_doors_sim(n=3, switch=True, num_trials=int(1e3)): 
#    winners = np.sum([make_a_deal(n, switch) for kk in range(num_trials)])
#    state = "switching" if switch else "not switching"
#    print("P(winning by "+state+" = {:.4f}".format(winners/num_trials))

With $n=3$ doors to check...

In [9]:
#n_doors_sim(n=3, switch=True, num_trials=int(1e5))
#n_doors_sim(n=3, switch=False, num_trials=int(1e5))

And now with $n=4$ or $n=5$ doors...

In [10]:
#print('\n--- with n=4 doors ---')
#n_doors_sim(n=4, switch=True, num_trials=int(1e5))
#n_doors_sim(n=4, switch=False, num_trials=int(1e5))

#print('\n--- with n=5 doors ---')
#n_doors_sim(n=5, switch=True, num_trials=int(1e5))
#n_doors_sim(n=5, switch=False, num_trials=int(1e5))

**Part B:** Verify your simulation results for the cases of $n=4$ and $n=5$ doors by hand.  What is the probability of winning if you switch, or do not switch, in those cases?

**Solution:**

$n=4$ doors case:

**Intuition:**  For the $n=4$ case, each door initially has 0.25 probability.  So the not-your-first choice doors share 0.75 probability.  Then Monty shows you that one of those three doors is a loser.  So, the 0.75 probability of the not-your-first-choice doors is now split evenly among the other two.  That means if you switch, you should find 0.75/2 = 0.375 probability of winning.  And your original door still has only 0.25 probability of winning if you do not switch.

**Math:**

Check it for yourself using Bayes' theorem and the Law of Total Probability!